### ISSUE TITLE
AI ShipDesign cache corruption

### ISSUE THREAD

**Issue by Dilvish-fo** [freeorion/freeorion#2092]
Upon savegame load, or sometimes in multiplayer games (perhaps only when using auto turn advance) the AI sometimes winds up with its ShipDesign records being corrupted for some reason.  There has long been some code in place to check for that and do some cleanup, logging a warning, but in at least some cases that cleanup has been incomplete and related errors still wind up being logged and sent to the players ingame (although the AI was still catching the errors and handling them smoothly).

I'm about to submit a PR that augments the cleanup and thus prevents the ingame error warning, but I think the remaining/underlying cache corruption problem still merits an Issue.

Below is a sample of the related logging.
```
14:33:01.392519 [info] python : ShipDesignAI.py:150 - ==========Updating ShipDesignCache for new turn==========
14:33:01.392519 [debug] python : ShipDesignAI.py:349 - Checking persistent cache for consistency...
14:33:01.393020 [warn] python : ShipDesignAI.py:370 - ShipID cache corrupted. Expected: EY Cerberus Mk. 5, got: AI_TESTDESIGN_PART_SH_DEFENSE_GRID_SH_COLONY_BASE.
14:33:01.395021 [debug] python : ShipDesignAI.py:477 - Updating Testdesigns for hulls...
14:33:01.395021 [debug] python : ShipDesignAI.py:490 - Caching buildable hulls per planet...
14:33:01.395522 [debug] python : ShipDesignAI.py:510 - Updating test designs for ship parts...
14:33:01.396522 [debug] python : ShipDesignAI.py:566 - Caching buildable ship parts per planet...
14:33:01.452562 [debug] python : ShipDesignAI.py:2208 - Shipdesign EY Cerberus Mk. 5 seems not to exist: Caching as invalid design.
14:33:01.454563 [error] python : ShipDesignAI.py:1038 - SH_ROBOTIC-AR_ZORTRIUM_PLATE-SH_DEFENSE_GRID-SR_WEAPON_1_1-SR_WEAPON_1_1-SR_WEAPON_1_1 maps to EY Cerberus Mk. 5 in Cache.map_reference_design_name. But the design seems not to exist...
14:33:01.454563 [error] python : ShipDesignAI.py:1038 - Traceback (most recent call last):
14:33:01.454563 [error] python : ShipDesignAI.py:1038 -   File "C:/Users\Public\Documents\Development\FreeOrion\default\python\AI\ShipDesignAI.py", line 1034, in add_design
14:33:01.454563 [error] python : ShipDesignAI.py:1038 -     return _get_design_by_name(Cache.map_reference_design_name[reference_name]).id
14:33:01.454563 [error] python : ShipDesignAI.py:1038 - AttributeError: 'NoneType' object has no attribute 'id'
14:33:01.455064 [error] python : ShipDesignAI.py:1187 - The best design for WarShipDesigner on planet 1918 could not be added.
14:33:01.484085 [error] python : ShipDesignAI.py:1038 - SH_ROBOTIC-AR_ZORTRIUM_PLATE-SH_DEFENSE_GRID-SR_WEAPON_1_1-SR_WEAPON_1_1-SR_WEAPON_1_1 maps to EY Cerberus Mk. 5 in Cache.map_reference_design_name. But the design seems not to exist...
14:33:01.484085 [error] python : ShipDesignAI.py:1038 - Traceback (most recent call last):
14:33:01.484085 [error] python : ShipDesignAI.py:1038 -   File "C:/Users\Public\Documents\Development\FreeOrion\default\python\AI\ShipDesignAI.py", line 1034, in add_design
14:33:01.484085 [error] python : ShipDesignAI.py:1038 -     return _get_design_by_name(Cache.map_reference_design_name[reference_name]).id
14:33:01.484085 [error] python : ShipDesignAI.py:1038 - AttributeError: 'NoneType' object has no attribute 'id'
14:33:01.485084 [error] python : ShipDesignAI.py:1187 - The best design for WarShipDesigner on planet 8302 could not be added.
```

**Comment by geoffthemedio** [comment]
Possibly this isn't limited to loading games? I set up a multiplayer game with 7 AIs, 200 stars, and let it auto-cycle turns. On turn 95, I started getting errors in the chat from the AI:
````
Begin Turn 95
Saving...
Saved 934609 bytes to file: C:\Users\Geoff\AppData\Roaming\FreeOrion\save\auto\FreeOrion__Observer_0095_20180511_144647.mps
[11 May 14:46:51] AI_7: AI_Error: AI script error : ShipDesignAI.py:add_design():1038  - SH_ASTEROID---AR_ZORTRIUM_PLATE-AR_ZORTRIUM_PLATE-AR_ZORTRIUM_PLATE-SR_WEAPON_2_1 maps to CN Cerberus Mk. 6 in Cache.map_reference_design_name. But the design seems not to exist...
Traceback (most recent call last):
  File "C:/Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion\default\python\AI\ShipDesignAI.py", line 1034, in add_design
    return _get_design_by_name(Cache.map_reference_design_name[reference_name]).id
AttributeError: 'NoneType' object has no attribute 'id'
[11 May 14:46:51] AI_7: AI_Error: AI script error : ShipDesignAI.py:optimize_design():1187  - The best design for WarShipDesigner on planet 424 could not be added.
[11 May 14:46:51] AI_7: AI_Error: AI script error : ShipDesignAI.py:add_design():1038  - SH_ASTEROID---AR_ZORTRIUM_PLATE-AR_ZORTRIUM_PLATE-AR_ZORTRIUM_PLATE-SR_WEAPON_2_1 maps to CN Cerberus Mk. 6 in Cache.map_reference_design_name. But the design seems not to exist...
Traceback (most recent call last):
  File "C:/Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion\default\python\AI\ShipDesignAI.py", line 1034, in add_design
    return _get_design_by_name(Cache.map_reference_design_name[reference_name]).id
AttributeError: 'NoneType' object has no attribute 'id'
[11 May 14:46:51] AI_7: AI_Error: AI script error : ShipDesignAI.py:optimize_design():1187  - The best design for WarShipDesigner on planet 2256 could not be added.
````

I loaded an autosave from turn 94, and got
````
[11 May 14:51:40] AI_5: AI_Error: AI script error : FleetUtilsAI.py:get_fleets_for_mission():151  - After having split a fleet, the original fleet apparently no longer exists.
Traceback (most recent call last):
  File "C:/Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion\default\python\AI\FleetUtilsAI.py", line 149, in get_fleets_for_mission
    fleet_pool_set.remove(fleet_id)
KeyError: 17495
````

But the original error didn't reproduce. A few turns later I did get this though:
````
Begin Turn 106
Saving...
Saved 1066064 bytes to file: C:\Users\Geoff\AppData\Roaming\FreeOrion\save\auto\FreeOrion__Observer_0106_20180511_145341.mps
[11 May 14:53:46] AI_2: AI_Error: AI script error : ShipDesignAI.py:add_design():1038  - SH_ROBOTIC--AR_ZORTRIUM_PLATE-AR_ZORTRIUM_PLATE-SR_WEAPON_2_1-SR_WEAPON_2_1 maps to FN Griffin Mk. 2 in Cache.map_reference_design_name. But the design seems not to exist...
Traceback (most recent call last):
  File "C:/Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion\default\python\AI\ShipDesignAI.py", line 1034, in add_design
    return _get_design_by_name(Cache.map_reference_design_name[reference_name]).id
AttributeError: 'NoneType' object has no attribute 'id'
[11 May 14:53:46] AI_2: AI_Error: AI script error : ShipDesignAI.py:optimize_design():1187  - The best design for WarShipDesigner on planet 1352 could not be added.
[11 May 14:53:46] AI_2: AI_Error: AI script error : ShipDesignAI.py:add_design():1038  - SH_ROBOTIC--AR_ZORTRIUM_PLATE-AR_ZORTRIUM_PLATE-SR_WEAPON_2_1-SR_WEAPON_2_1 maps to FN Griffin Mk. 2 in Cache.map_reference_design_name. But the design seems not to exist...
Traceback (most recent call last):
  File "C:/Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion\default\python\AI\ShipDesignAI.py", line 1034, in add_design
    return _get_design_by_name(Cache.map_reference_design_name[reference_name]).id
AttributeError: 'NoneType' object has no attribute 'id'
[11 May 14:53:46] AI_2: AI_Error: AI script error : ShipDesignAI.py:optimize_design():1187  - The best design for WarShipDesigner on planet 1208 could not be added.
[11 May 14:53:46] AI_2: AI_Error: AI script error : ShipDesignAI.py:add_design():1038  - SH_ROBOTIC--AR_ZORTRIUM_PLATE-AR_ZORTRIUM_PLATE-SR_WEAPON_2_1-SR_WEAPON_2_1 maps to FN Griffin Mk. 2 in Cache.map_reference_design_name. But the design seems not to exist...
Traceback (most recent call last):
  File "C:/Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion\default\python\AI\ShipDesignAI.py", line 1034, in add_design
    return _get_design_by_name(Cache.map_reference_design_name[reference_name]).id
AttributeError: 'NoneType' object has no attribute 'id'
[11 May 14:53:46] AI_2: AI_Error: AI script error : ShipDesignAI.py:optimize_design():1187  - The best design for WarShipDesigner on planet 296 could not be added.
[11 May 14:53:46] AI_2: AI_Error: AI script error : ShipDesignAI.py:add_design():1038  - SH_ROBOTIC--AR_ZORTRIUM_PLATE-AR_ZORTRIUM_PLATE-SR_WEAPON_2_1-SR_WEAPON_2_1 maps to FN Griffin Mk. 2 in Cache.map_reference_design_name. But the design seems not to exist...
Traceback (most recent call last):
  File "C:/Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion\default\python\AI\ShipDesignAI.py", line 1034, in add_design
    return _get_design_by_name(Cache.map_reference_design_name[reference_name]).id
AttributeError: 'NoneType' object has no attribute 'id'
[11 May 14:53:46] AI_2: AI_Error: AI script error : ShipDesignAI.py:optimize_design():1187  - The best design for WarShipDesigner on planet 1208 could not be added.
[11 May 14:53:46] AI_2: AI_Error: AI script error : ShipDesignAI.py:add_design():1038  - SH_ROBOTIC--AR_ZORTRIUM_PLATE-AR_ZORTRIUM_PLATE-SR_WEAPON_2_1-SR_WEAPON_2_1 maps to FN Griffin Mk. 2 in Cache.map_reference_design_name. But the design seems not to exist...
Traceback (most recent call last):
  File "C:/Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion\default\python\AI\ShipDesignAI.py", line 1034, in add_design
    return _get_design_by_name(Cache.map_reference_design_name[reference_name]).id
AttributeError: 'NoneType' object has no attribute 'id'
[11 May 14:53:46] AI_2: AI_Error: AI script error : ShipDesignAI.py:optimize_design():1187  - The best design for WarShipDesigner on planet 296 could not be added.
[11 May 14:53:46] AI_2: AI_Error: AI script error : ShipDesignAI.py:add_design():1038  - SH_ROBOTIC--AR_ZORTRIUM_PLATE-AR_ZORTRIUM_PLATE-SR_WEAPON_2_1-SR_WEAPON_2_1 maps to FN Griffin Mk. 2 in Cache.map_reference_design_name. But the design seems not to exist...
Traceback (most recent call last):
  File "C:/Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion\default\python\AI\ShipDesignAI.py", line 1034, in add_design
    return _get_design_by_name(Cache.map_reference_design_name[reference_name]).id
AttributeError: 'NoneType' object has no attribute 'id'
[11 May 14:53:46] AI_2: AI_Error: AI script error : ShipDesignAI.py:optimize_design():1187  - The best design for WarShipDesigner on planet 1208 could not be added.
[11 May 14:53:46] AI_2: AI_Error: AI script error : ShipDesignAI.py:add_design():1038  - SH_ROBOTIC--AR_ZORTRIUM_PLATE-AR_ZORTRIUM_PLATE-SR_WEAPON_2_1-SR_WEAPON_2_1 maps to FN Griffin Mk. 2 in Cache.map_reference_design_name. But the design seems not to exist...
Traceback (most recent call last):
  File "C:/Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion\default\python\AI\ShipDesignAI.py", line 1034, in add_design
    return _get_design_by_name(Cache.map_reference_design_name[reference_name]).id
AttributeError: 'NoneType' object has no attribute 'id'
[11 May 14:53:46] AI_2: AI_Error: AI script error : ShipDesignAI.py:optimize_design():1187  - The best design for WarShipDesigner on planet 296 could not be added.
[11 May 14:53:46] AI_2: AI_Error: AI script error : ShipDesignAI.py:add_design():1038  - SH_ROBOTIC--AR_ZORTRIUM_PLATE-AR_ZORTRIUM_PLATE-SR_WEAPON_2_1-SR_WEAPON_2_1 maps to FN Griffin Mk. 2 in Cache.map_reference_design_name. But the design seems not to exist...
Traceback (most recent call last):
  File "C:/Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion\default\python\AI\ShipDesignAI.py", line 1034, in add_design
    return _get_design_by_name(Cache.map_reference_design_name[reference_name]).id
AttributeError: 'NoneType' object has no attribute 'id'
[11 May 14:53:46] AI_2: AI_Error: AI script error : ShipDesignAI.py:optimize_design():1187  - The best design for WarShipDesigner on planet 1208 could not be added.
[11 May 14:53:46] AI_2: AI_Error: AI script error : ShipDesignAI.py:add_design():1038  - SH_ROBOTIC--AR_ZORTRIUM_PLATE-AR_ZORTRIUM_PLATE-SR_WEAPON_2_1-SR_WEAPON_2_1 maps to FN Griffin Mk. 2 in Cache.map_reference_design_name. But the design seems not to exist...
Traceback (most recent call last):
  File "C:/Users\Geoff\Desktop\FreeOrionSDK5\FreeOrion\default\python\AI\ShipDesignAI.py", line 1034, in add_design
    return _get_design_by_name(Cache.map_reference_design_name[reference_name]).id
AttributeError: 'NoneType' object has no attribute 'id'
[11 May 14:53:46] AI_2: AI_Error: AI script error : ShipDesignAI.py:optimize_design():1187  - The best design for WarShipDesigner on planet 296 could not be added.
````

**Comment by Dilvish-fo** [comment]
> Possibly this isn't limited to loading games? I set up a multiplayer game with 7 AIs, 200 stars, and let it auto-cycle turns

Ah, right-- I had added that in the description because that's how I had triggered it and one or more of the functions I was dealing with were specifically directed to dealing with the cache after game load, but I also see that the general method for verifying the cache has this comment from Morlic:
```
Especially (only?) in multiplayer games, the shipDesignIDs may sometimes change across turns.
```
And it's now ringing a bell that I had brought this to his attention after doing the same thing you just did-- running a game for a number of turns  auto-advancing.  I don't recall just what he did to figure out that the shipDesignIDs were actually changing.  I suppose we should re-verify that and then probably open a new Issue focused more specifically on that.

I take it that you got these errors without the recent PR I just put up?  It doesn't solve the underlying problem but does some more AI cleanup and I think would have prevented the error spam to user.

**Comment by Morlic-fo** [comment]
> I don't recall just what he did to figure out that the shipDesignIDs were actually changing.

I think this is the relevant forum post:
http://www.freeorion.org/forum/viewtopic.php?f=9&t=9296&start=30#p75416

TL;DR: Run through all ids in empire.allShipDesigns and compare server-queried results with cache from last turn.

**Comment by Morlic-fo** [comment]
> Upon savegame load (...) the AI sometimes winds up with its ShipDesign records being corrupted for some reason.

Does this happen immediately on savegame load or some later turn?

If it happens immediately on load, that is both weird and quite an important hint. The cache is not part of the savegame state but rebuilt from server queries (and then immediately checked for corruption).

Relevant code
```
def update_for_new_turn(self):
        [...]
        if not self.map_reference_design_name:
            self._build_cache_after_load()
        self._check_cache_for_consistency()
        [...]

def _build_cache_after_load(self):
        """Build cache after loading or starting a game.

        This function is supposed to be called after a reload or at the first turn.
        It reads out all the existing ship designs and then updates the following cache:
        - map_reference_design_name
        - design_id_by_name
        """
        if self.map_reference_design_name or self.design_id_by_name:
            warn("ShipDesignAI.Cache._build_cache_after_load() called but cache is not empty.")
        for design_id in fo.getEmpire().allShipDesigns:
            design = fo.getShipDesign(design_id)
            if TESTDESIGN_NAME_BASE in design.name:
                continue
            reference_name = _build_reference_name(design.hull, design.parts)
            self.map_reference_design_name[reference_name] = design.name
            self.design_id_by_name[design.name] = design_id
```

**Comment by Vezzra** [comment]
@freeorion/ai-team, status on this? Any progress?

**Comment by geoffthemedio** [comment]
@freeorion/ai-team poke
